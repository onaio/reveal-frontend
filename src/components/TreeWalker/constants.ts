export const ACTIVE = 'Active';
export const FEATURE = 'Feature';
export const FIND_BY_ID = 'location/findByJurisdictionIds';
export const FIND_BY_PROPERTIES = 'location/findByProperties';
export const LOCATION = 'location';
export const COULDNT_LOAD_CHILDREN = 'Could not load children';
export const COULDNT_LOAD_PARENTS = 'Could not load parents';
export const NOT_AVAILABLE = 'N/A';
