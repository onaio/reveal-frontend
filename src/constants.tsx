// strings
export const MAP = 'map';
export type MAP = typeof MAP;
export const LOCATION = 'Location';
export type LOCATION = typeof LOCATION;
export const PROVINCE = 'Province';
export type PROVINCE = typeof PROVINCE;
export const HOME = 'Home';
export type HOME = typeof HOME;
export const FOCUS_INVESTIGATION = 'Focus Investigation';
export type FOCUS_INVESTIGATION = typeof FOCUS_INVESTIGATION;
export const ACTIVE_FOCUS_INVESTIGATION = 'Active Focus Investigations';
export type ACTIVE_FOCUS_INVESTIGATION = typeof ACTIVE_FOCUS_INVESTIGATION;
export const FOCUS_AREA_INFO = 'Focus Area Information';
export type FOCUS_AREA_INFO = typeof FOCUS_AREA_INFO;
export const FOCUS_INVESTIGATIONS = 'Focus Investigations';
export type FOCUS_INVESTIGATIONS = typeof FOCUS_INVESTIGATIONS;
export const MAP_ID = 'map-1';
export type MAP_ID = typeof MAP_ID;
export const GEOJSON = 'geojson';
export type GEOJSON = typeof GEOJSON;
export const STRINGIFIED_GEOJSON = 'stringified-geojson';
export type STRINGIFIED_GEOJSON = typeof STRINGIFIED_GEOJSON;
export const FOCUS_AREA_HEADER = 'Focus Area';
export type FOCUS_AREA_HEADER = typeof FOCUS_AREA_HEADER;
export const REASON_HEADER = 'Reason';
export type REASON_HEADER = typeof REASON_HEADER;
export const STATUS_HEADER = 'Status';
export type STATUS_HEADER = typeof STATUS_HEADER;
export const CASE_NOTIF_DATE_HEADER = 'Case Notif. Date';
export type CASE_NOTIF_DATE_HEADER = typeof CASE_NOTIF_DATE_HEADER;
export const CASE_CLASSIFICATION_HEADER = 'Case Class.';
export type CASE_CLASSIFICATION_HEADER = typeof CASE_CLASSIFICATION_HEADER;
export const DEFINITIONS = 'Definitions';
export type DEFINITIONS = typeof DEFINITIONS;
export const IN = 'in';
export type IN = typeof IN;
export const DISTRICT = 'District';
export type DISTRICT = typeof DISTRICT;
export const CANTON = 'Canton';
export type CANTON = typeof CANTON;
export const FI_REASON = 'FI Reason';
export type FI_REASON = typeof FI_REASON;
export const FI_STATUS = 'FI Status';
export type FI_STATUS = typeof FI_STATUS;
export const ACTIVE_INVESTIGATION = 'Active Investigation';
export type ACTIVE_INVESTIGATION = typeof ACTIVE_INVESTIGATION;
export const COMPLETE = 'Complete';
export type COMPLETE = typeof COMPLETE;
export const NO = 'No';
export type NO = typeof NO;
export const RESPONSE = 'Response';
export type RESPONSE = typeof RESPONSE;
export const MEASURE = 'Measure';
export type MEASURE = typeof MEASURE;
export const OF = 'of';
export type OF = typeof OF;
export const MARK_AS_COMPLETE = 'Mark as complete';
export type MARK_AS_COMPLETE = typeof MARK_AS_COMPLETE;
export const TARGET = 'Target';
export type TARGET = typeof TARGET;
export const POLYGON = 'Polygon';
export type POLYGON = typeof POLYGON;
export const MULTI_POLYGON = 'MultiPolygon';
export type MULTI_POLYGON = typeof MULTI_POLYGON;
export const POINT = 'Point';
export type POINT = typeof POINT;
export const FEATURE = 'Feature';
export type FEATURE = typeof FEATURE;
export const FEATURE_COLLECTION = 'FeatureCollection';
export type FEATURE_COLLECTION = typeof FEATURE_COLLECTION;
export const MAIN_PLAN = 'main-plan-layer';
export type MAIN_PLAN = typeof MAIN_PLAN;
export const APP = 'APP';
export type APP = typeof APP;
export const NO_GEOMETRIES_RESPONSE = 'Goals have no Geometries';
export type NO_GEOMETRIES_RESPONSE = typeof NO_GEOMETRIES_RESPONSE;

// internal urls
export const LOGIN_URL = '/login';
export type LOGIN_URL = typeof LOGIN_URL;
export const LOGOUT_URL = '/logout';
export type LOGOUT_URL = typeof LOGOUT_URL;
export const HOME_URL = '/';
export type HOME_URL = typeof HOME_URL;
export const IRS_URL = '/irs';
export type IRS_URL = typeof IRS_URL;
export const FI_URL = '/focus-investigation';
export type FI_URL = typeof FI_URL;
export const FI_SINGLE_URL = '/focus-investigation/view';
export type FI_SINGLE_URL = typeof FI_SINGLE_URL;
export const FI_SINGLE_MAP_URL = '/focus-investigation/map';
export type FI_SINGLE_MAP_URL = typeof FI_SINGLE_MAP_URL;
export const FI_HISTORICAL_URL = '/focus-investigation/historical';
export type FI_HISTORICAL_URL = typeof FI_HISTORICAL_URL;

// colors
export const GREEN = 'Green';
export type GREEN = typeof GREEN;
export const YELLOW = 'Yellow';
export type YELLOW = typeof YELLOW;
export const ORANGE = 'Orange';
export type ORANGE = typeof ORANGE;
export const RED = 'Red';
export type RED = typeof RED;

// task action codes
export const RACD_REGISTER_FAMILY_CODE = 'RACD Register Family';
export type RACD_REGISTER_FAMILY_CODE = typeof RACD_REGISTER_FAMILY_CODE;
export const CASE_CONFIRMATION_CODE = 'Case Confirmation';
export type CASE_CONFIRMATION_CODE = typeof CASE_CONFIRMATION_CODE;
export const MOSQUITO_COLLECTION_CODE = 'Mosquito Collection';
export type MOSQUITO_COLLECTION_CODE = typeof MOSQUITO_COLLECTION_CODE;
export const BEDNET_DISTRIBUTION_CODE = 'Bednet Distribution';
export type BEDNET_DISTRIBUTION_CODE = typeof BEDNET_DISTRIBUTION_CODE;
export const LARVAL_DIPPING_CODE = 'Larval Dipping';
export type LARVAL_DIPPING_CODE = typeof LARVAL_DIPPING_CODE;
export const BLOOD_SCREENING_CODE = 'Blood Screening';
export type BLOOD_SCREENING_CODE = typeof BLOOD_SCREENING_CODE;
export const IRS_CODE = 'IRS';
export type IRS = typeof IRS_CODE;
