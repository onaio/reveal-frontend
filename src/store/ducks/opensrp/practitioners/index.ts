/** practitioners ducks modules: actions, actionCreators, reducer and selectors */
import { get, keyBy, keys, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'practitioner';

/** Interface for practitioner json object */
export interface Practitioner {
  active: true;
  identifier: string;
  name: string;
  userId: string;
  username: string;
}

/** interface of practitionerRole - relation: practitioners belonging to which organization */
export interface PractitionerRole {
  [organizationId: string]: { [practitionerId: string]: Practitioner };
}

// actions

/** action type for fetching practitioners */
export const PRACTITIONERS_FETCHED = 'opensrp/reducer/practitioners/PRACTITIONERS_FETCHED';
/** action type for removing practitioners */
export const REMOVE_PRACTITIONERS = 'opensrp/reducer/practitioners/REMOVE_PRACTITIONERS';
/** action type for fetching practitionerRoles - information as to what organization a practitioner belongs to */
export const PRACTITIONER_ROLES_FETCHED =
  'opensrp/reducer/practitioners/PRACTITIONER_ROLES_FETCHED';
/** action type for removing practitionerRoles */
export const REMOVE_PRACTITIONER_ROLES = 'opensrp/reducer/practitioners/REMOVE_PRACTITIONER_ROLES';

/** interface action to add practitioners to store */
export interface FetchPractitionersAction extends AnyAction {
  practitionersById: { [key: string]: Practitioner };
  type: typeof PRACTITIONERS_FETCHED;
}

/** Interface for removePractitionersAction */
interface RemovePractitionersAction extends AnyAction {
  practitionersById: {};
  type: typeof REMOVE_PRACTITIONERS;
}

/** interface for actions that add practitionerRoles */
interface FetchPractitionerRolesAction extends AnyAction {
  practitionerRoles: PractitionerRole;
  practitionersById: { [key: string]: Practitioner };
  type: typeof PRACTITIONER_ROLES_FETCHED;
}

/** interface for actions that remove PractitionerRoles */
interface RemovePractitionerRolesAction extends AnyAction {
  practitionerRoles: {};
  type: typeof REMOVE_PRACTITIONER_ROLES;
}

/** Create type for practitioners reducer actions */
export type PractitionersActionTypes =
  | FetchPractitionersAction
  | RemovePractitionersAction
  | FetchPractitionerRolesAction
  | RemovePractitionerRolesAction
  | AnyAction;

// action Creators

/** Fetch practitioners action creator
 * @param {Practitioner []} practitionersList - practitioners array to add to store
 * @return {FetchPractitionersAction} - an action to add practitioners to redux store
 */
export const fetchPractitioners = (
  practitionersList: Practitioner[] = []
): FetchPractitionersAction => ({
  practitionersById: keyBy(
    practitionersList,
    (practitioner: Practitioner) => practitioner.identifier
  ),
  type: PRACTITIONERS_FETCHED,
});

// actions

/** removePractitionersAction action */
export const removePractitionersAction = {
  practitionersById: {},
  type: REMOVE_PRACTITIONERS,
};

/** action creator for adding info on organization to practitioners
 * this will modify both the practitionersById as well as the practitionerRoles
 * Use this action creator when pulling data for practitioners belonging to
 * an organization only, otherwise see fetchPractitioners
 *
 * @param {Practitioner []} practitioners - an array of practitioners
 * @param {string}  organizationId - id of organization that the practitioners belong to
 */
export const fetchPractitionerRoles = (
  practitioners: Practitioner[],
  organizationId: string
): FetchPractitionerRolesAction => {
  const practitionersById = keyBy(
    practitioners,
    (practitioner: Practitioner) => practitioner.identifier
  );

  return {
    practitionerRoles: { [organizationId]: practitionersById },
    practitionersById,
    type: PRACTITIONER_ROLES_FETCHED,
  };
};

/** Action to remove all PractitionerRoles from store */
export const removePractitionerRolesAction: RemovePractitionerRolesAction = {
  practitionerRoles: {},
  type: REMOVE_PRACTITIONER_ROLES,
};

// The reducer

/** interface for practitioners state in redux store */
interface PractitionerState {
  practitionersById: { [key: string]: Practitioner };
  practitionerRoles: PractitionerRole;
}

/** Create an immutable practitioners state */
export type ImmutablePractitionersState = PractitionerState &
  SeamlessImmutable.ImmutableObject<PractitionerState>;

/** initial practitioners-state state */
export const initialState: ImmutablePractitionersState = SeamlessImmutable({
  practitionerRoles: {},
  practitionersById: {},
});

/** the practitioners reducer function */
export default function reducer(
  state: ImmutablePractitionersState = initialState,
  action: PractitionersActionTypes
): ImmutablePractitionersState {
  switch (action.type) {
    case PRACTITIONERS_FETCHED:
      return SeamlessImmutable({
        ...state,
        practitionersById: { ...state.practitionersById, ...action.practitionersById },
      });
    case REMOVE_PRACTITIONERS:
      return SeamlessImmutable({
        ...state,
        practitionersById: action.practitionersById,
      });
    case PRACTITIONER_ROLES_FETCHED:
      const organizationId = keys(action.practitionerRoles)[0];
      return SeamlessImmutable({
        ...state,
        practitionerRoles: {
          ...state.practitionerRoles,
          [organizationId]: {
            ...state.practitionerRoles[organizationId],
            ...action.practitionerRoles[organizationId],
          },
        } as PractitionerRole & SeamlessImmutable.ImmutableObject<PractitionerRole>,
        practitionersById: { ...state.practitionersById, ...action.practitionersById },
      });
    case REMOVE_PRACTITIONER_ROLES:
      return SeamlessImmutable({
        ...state,
        practitionerRoles: action.practitionerRoles,
      });
    default:
      return state;
  }
}

// Selectors

/** returns all practitioners in the store as values whose keys are their respective ids
 * @param {Partial<Store>} state - the redux store
 * @return { { [key: string] : Practitioner} } - practitioners object as values, respective ids as keys
 */
export function getPractitionersById(state: Partial<Store>): { [key: string]: Practitioner } {
  return (state as any)[reducerName].practitionersById;
}

/** gets practitioners as an array of practitioners objects
 * @param {Partial<Store>} state - the redux store
 * @return {Practitioner[]} - an array of practitioners objs
 */
export function getPractitionersArray(state: Partial<Store>): Practitioner[] {
  return values(getPractitionersById(state));
}

/** get a specific practitioner by their id
 * @param {Partial<Store>} state - the redux store
 * @return {Practitioner | null} a practitioner obj if the id is found else null
 */
export function getPractitionerById(state: Partial<Store>, id: string): Practitioner | null {
  return get(getPractitionersById(state), id) || null;
}

/** get practitioners that belong to a certain organization
 * @param {Partial<Store>} state -  the redux store
 * @param {string} organizationId - get practitioners that belong to organization with this id
 *
 * @return {Practitioner[]} - returns a list of practitioners that belong to organization with given id
 */
export function getPractitionersByOrgId(
  state: Partial<Store>,
  organizationId: string
): Practitioner[] {
  let practitionersById = (state as any)[reducerName].practitionerRoles[organizationId];
  practitionersById = practitionersById !== undefined ? practitionersById : {};
  return values(practitionersById);
}
