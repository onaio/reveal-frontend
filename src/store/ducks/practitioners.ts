/** practitioners ducks modules: actions, actionCreators, reducer and selectors */

import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'practitioners';

/** Interface for practitioner json object */
export interface Practitioner {
  active: true;
  identifier: string;
  name: string;
  userId: string;
  username: string;
}

// actions

/** action type for fetching practitioners */
export const PRACTITIONERS_FETCHED = 'opensrp/reducer/practitioners/PRACTITIONERS_FETCHED';
/** action type for removing practitioners */
export const REMOVE_PRACTITIONERS = 'opensrp/reducer/practitioners/REMOVE_PRACTITIONERS';

/** interface action to add practitioners to store */
export interface FetchPractitionersAction extends AnyAction {
  practitionersById: { [key: string]: Practitioner };
  type: typeof PRACTITIONERS_FETCHED;
}

/** Interface for removePractitionersAction */
interface RemovePractitionersAction extends AnyAction {
  practitionersById: {};
  type: typeof REMOVE_PRACTITIONERS;
}

/** Create type for practitioners reducer actions */
export type PractitionersActionTypes =
  | FetchPractitionersAction
  | RemovePractitionersAction
  | AnyAction;

// action Creators

/** Fetch practitioners action creator
 * @param {Practitioner []} practitionersList - practitioners array to add to store
 * @return {FetchPractitionersAction} - an action to add practitioners to redux store
 */
export const fetchPractitioners = (
  practitionersList: Practitioner[] = []
): FetchPractitionersAction => ({
  practitionersById: keyBy(
    practitionersList,
    (practitioner: Practitioner) => practitioner.identifier
  ),
  type: PRACTITIONERS_FETCHED,
});

// actions

/** removePractitionersAction action */
export const removePractitionersAction = {
  practitionersById: {},
  type: REMOVE_PRACTITIONERS,
};

// The reducer

/** interface for practitioners state in redux store */
interface PractitionerState {
  practitionersById: { [key: string]: Practitioner };
}

/** Create an immutable practitioners state */
export type ImmutablePractitionersState = PractitionerState &
  SeamlessImmutable.ImmutableObject<PractitionerState>;

/** initial practitioners-state state */
export const initialState: ImmutablePractitionersState = SeamlessImmutable({
  practitionersById: {},
});

/** the practitioners reducer function */
export default function reducer(
  state: ImmutablePractitionersState = initialState,
  action: PractitionersActionTypes
): ImmutablePractitionersState {
  switch (action.type) {
    case PRACTITIONERS_FETCHED:
      return SeamlessImmutable({
        ...state,
        practitionersById: { ...state.practitionersById, ...action.practitionersById },
      });
    case REMOVE_PRACTITIONERS:
      return SeamlessImmutable({
        ...state,
        practitionersById: action.practitionersById,
      });
    default:
      return state;
  }
}

// Selectors

/** returns all practitioners in the store as values whose keys are their respective ids
 * @param {Partial<Store>} state - the redux store
 * @return { { [key: string] : Practitioner} } - practitioners object as values, respective ids as keys
 */
export function getPractitionersById(state: Partial<Store>): { [key: string]: Practitioner } {
  return (state as any)[reducerName].practitionersById;
}

/** gets practitioners as an array of practitioners objects
 * @param {Partial<Store>} state - the redux store
 * @return {Practitioner[]} - an array of practitioners objs
 */
export function getPractitionersArray(state: Partial<Store>): Practitioner[] {
  return values(getPractitionersById(state));
}

/** get a specific practitioner by their id
 * @param {Partial<Store>} state - the redux store
 * @return {Practitioner | null} a practitioner obj if the id is found else null
 */
export function getPractitionerById(state: Partial<Store>, id: string): Practitioner | null {
  return get(getPractitionersById(state), id) || null;
}
