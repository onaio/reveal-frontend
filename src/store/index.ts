import { store } from '@opensrp/store';

export default store;
