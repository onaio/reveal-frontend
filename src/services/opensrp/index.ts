import { IncomingHttpHeaders } from 'http';
import { OPENSRP_API_BASE_URL } from '../../configs/env';
import store from '../../store';
import { getAccessToken } from '../../store/selectors';

/** allowed http methods */
type HTTPMethod = 'GET' | 'POST' | 'PUT';

/** get default HTTP headers for OpenSRP service */
export function getDefaultHeaders(
  accept: string = 'application/json',
  authorizationType: string = 'Bearer',
  contentType: string = 'application/json;charset=UTF-8'
): IncomingHttpHeaders {
  return {
    accept,
    authorization: `${authorizationType} ${getAccessToken(store.getState())}`,
    'content-type': contentType,
  };
}

/** interface to describe URL params object */
export interface URLParams {
  [key: string]: string | number | boolean;
}

/** params option type */
type paramsType = URLParams | null;

/** converts URL params object to string
 * @param {URLParams} obj - the object representing URL params
 * @returns {string} URL params as a string
 */
export function getURLParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}=${val}`)
    .join('&');
}

/** converts filter params object to string
 * @param {URLParams} obj - the object representing filter params
 * @returns {string} filter params as a string
 */
export function getFilterParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}:${val}`)
    .join(',');
}

/** get payload for fetch
 * @param {HTTPMethod} method - the HTTP method
 * @returns the payload
 */
export function getPayload(method: HTTPMethod) {
  return {
    headers: getDefaultHeaders() as HeadersInit,
    method,
  };
}

/** Get URL
 * @param {string} url - the url
 * @param {paramType} params - the url params object
 * @returns {string} the final url
 */
export function getURL(url: string, params: paramsType = null): string {
  let result = url;
  if (params) {
    result = `${result}?${getURLParams(params)}`;
  }
  return result;
}

/** The OpenSRP service class
 *
 * Sample usage:
 * -------------
 * const service = new OpenSRPService('the-endpoint');
 *
 * **To list all objects**: service.list()
 *
 * **To get one object**: service.read('the-object-identifier')
 *
 * **To create a new object**: service.create(theObject)
 *
 * **To update an object**: service.update(theObject)
 */
export class OpenSRPService {
  public baseURL: string;
  public endpoint: string;
  public generalURL: string;

  constructor(endpoint: string, baseURL: string = OPENSRP_API_BASE_URL) {
    this.endpoint = endpoint;
    this.baseURL = baseURL;
    this.generalURL = `${this.baseURL}${this.endpoint}`;
  }

  /** create method
   * Send a POST request to the general endpoint containing the new object data
   * Successful requests will result in a HTTP status 201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async create<T>(data: T, params: paramsType = null, method: HTTPMethod = 'POST') {
    const url = getURL(this.generalURL, params);
    const payload = {
      ...getPayload(method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok || response.status !== 201) {
      throw new Error(
        `OpenSRPService create on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }

  /** read method
   * Send a GET request to the url for the specific object
   * @param {string|number} id - the identifier of the object
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async read(id: string | number, params: paramsType = null, method: HTTPMethod = 'GET') {
    const url = getURL(`${this.generalURL}/${id}`, params);
    const response = await fetch(url, getPayload(method));

    if (!response.ok) {
      throw new Error(
        `OpenSRPService read on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return await response.json();
  }

  /** update method
   * Simply send the updated object as PUT request to the general endpoint URL
   * Successful requests will result in a HTTP status 200/201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async update<T>(data: T, params: paramsType = null, method: HTTPMethod = 'PUT') {
    const url = getURL(this.generalURL, params);
    const payload = {
      ...getPayload(method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok) {
      throw new Error(
        `OpenSRPService update on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }

  /** list method
   * Send a GET request to the general API endpoint
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns list of objects returned by API
   */
  public async list(params: paramsType = null, method: HTTPMethod = 'GET') {
    const url = getURL(this.generalURL, params);
    const response = await fetch(url, getPayload(method));

    if (!response.ok) {
      throw new Error(
        `OpenSRPService list on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return await response.json();
  }
}
