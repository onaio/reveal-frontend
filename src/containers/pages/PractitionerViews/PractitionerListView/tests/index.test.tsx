import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { act } from 'react-dom/test-utils';
import Helmet from 'react-helmet';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import ConnectedPractitionersListView, { PractitionersListView } from '..';
import { NO_ROWS_FOUND, PRACTITIONERS } from '../../../../../configs/lang';
import { OPENSRP_PRACTITIONER_ENDPOINT, PRACTITIONERS_LIST_URL } from '../../../../../constants';
import store from '../../../../../store';
import * as practitionerDucks from '../../../../../store/ducks/opensrp/practitioners';
import * as fixtures from '../../../../../store/ducks/tests/fixtures';

// tslint:disable-next-line: no-var-requires
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

describe('src/containers/TeamAssignment/PractitionersListView/', () => {
  beforeEach(() => {
    store.dispatch(practitionerDucks.removePractitionersAction);
    jest.resetAllMocks();
  });

  it('renders a dumb TeamListView correctly', () => {
    fetch.once(JSON.stringify(fixtures.practitioners));
    const mock: any = jest.fn();
    const props = {
      history,
      location: mock,
      match: {
        isExact: true,
        params: { id: '' },
        path: PRACTITIONERS_LIST_URL,
        url: PRACTITIONERS_LIST_URL,
      },
      practitioners: fixtures.practitioners,
    };
    const wrapper = mount(
      <Router history={history}>
        <PractitionersListView {...props} />
      </Router>
    );
    // page title
    const helmet = Helmet.peek();
    expect(helmet.title).toEqual(`${PRACTITIONERS}(${fixtures.practitioners.length})`);

    // should display a breadcrumb
    expect(toJson(wrapper.find('Breadcrumb'))).toMatchSnapshot('Breadcrumb');

    // should have link to add practitioner
    expect(wrapper.find(`LinkAsButton`).length).toEqual(1);

    // should have a table
    expect(toJson(wrapper.find('tbody tr').first())).toMatchSnapshot(
      'First table record in listview'
    );
    wrapper.unmount();
  });

  it('Makes the expected calls to the opensrpService', async () => {
    const mockList = jest.fn(async () => fixtures.practitioners);
    const classMock = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    const fetchedMock = jest.fn(arg => practitionerDucks.fetchPractitioners(arg));
    const mock: any = jest.fn();
    const props = {
      fetchPractitionersCreator: fetchedMock,
      history,
      location: mock,
      match: {
        isExact: true,
        params: { id: '' },
        path: PRACTITIONERS_LIST_URL,
        url: PRACTITIONERS_LIST_URL,
      },
      practitioners: fixtures.practitioners,
      serviceClass: classMock,
    };
    mount(
      <Router history={history}>
        <PractitionersListView {...props} />
      </Router>
    );

    await flushPromises();
    expect(classMock).toBeCalledWith(OPENSRP_PRACTITIONER_ENDPOINT);
    expect(mockList).toHaveBeenCalled();
    expect(fetchedMock).toHaveBeenCalledWith(fixtures.practitioners, true);
  });

  it('Does not load when api returns no data', async () => {
    fetch.once(JSON.stringify([]));
    const mock: any = jest.fn();
    const props = {
      history,
      location: mock,
      match: {
        isExact: true,
        params: {},
        path: PRACTITIONERS_LIST_URL,
        url: PRACTITIONERS_LIST_URL,
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPractitionersListView {...props} />
        </Router>
      </Provider>
    );
    wrapper.update();

    expect(wrapper.find('Ripple').length).toEqual(1);

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchInlineSnapshot(
      `"HomePractitionersPractitioners (0)Add PractitionerPractitioner NameUsernameIdentifierActionsNo rows found"`
    );
    expect(wrapper.text().includes(NO_ROWS_FOUND)).toBeTruthy();
  });

  it('PractitionerListView works correctly when connected to store', () => {
    fetch.once(JSON.stringify([]));
    const mock: any = jest.fn();
    store.dispatch(practitionerDucks.fetchPractitioners(fixtures.practitioners));
    const props = {
      history,
      location: mock,
      match: {
        isExact: true,
        params: {},
        path: PRACTITIONERS_LIST_URL,
        url: PRACTITIONERS_LIST_URL,
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPractitionersListView {...props} />
        </Router>
      </Provider>
    );

    // check that store data is part of passed props
    const foundProps = wrapper.find('PractitionersListView').props() as any;
    expect(foundProps.practitioners).toEqual(fixtures.practitioners);
    wrapper.unmount();
  });

  it('search works correctly', () => {
    fetch.once(JSON.stringify([]));
    store.dispatch(practitionerDucks.fetchPractitioners(fixtures.practitioners));
    const props = {
      history,
      location: {
        pathname: PRACTITIONERS_LIST_URL,
        search: '?title=Biophics',
      },
      match: {
        isExact: true,
        params: {},
        path: PRACTITIONERS_LIST_URL,
        url: PRACTITIONERS_LIST_URL,
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPractitionersListView {...props} />
        </Router>
      </Provider>
    );

    // check that store data is part of passed props
    const foundProps = wrapper.find('PractitionersListView').props() as any;
    expect(foundProps.practitioners).toEqual([fixtures.practitioner2]);
    wrapper.unmount();
  });
});
