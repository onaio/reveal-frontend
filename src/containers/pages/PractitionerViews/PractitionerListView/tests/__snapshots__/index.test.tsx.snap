// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/containers/TeamAssignment/PractitionersListView/ renders a dumb TeamListView correctly: Breadcrumb 1`] = `
<Breadcrumb
  aria-label="breadcrumb"
  className="reveal-breadcrumb"
  listTag="ol"
  tag="nav"
>
  <nav
    aria-label="breadcrumb"
    className="reveal-breadcrumb"
  >
    <ol
      className="breadcrumb"
    >
      <BreadcrumbItem
        key="0"
        tag="li"
      >
        <li
          className="breadcrumb-item"
        >
          <Link
            key="0"
            to="/"
          >
            <LinkAnchor
              href="/"
              navigate={[Function]}
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Home
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </BreadcrumbItem>
      <BreadcrumbItem
        active={true}
        tag="li"
      >
        <li
          aria-current="page"
          className="active breadcrumb-item"
        >
          Practitioners
        </li>
      </BreadcrumbItem>
    </ol>
  </nav>
</Breadcrumb>
`;

exports[`src/containers/TeamAssignment/PractitionersListView/ renders a dumb TeamListView correctly: First table record in listview 1`] = `
<tr
  className="listview-tr"
  key="0"
>
  <ElementMap
    HTMLTag="td"
    className="listview-th"
    items={
      Array [
        "tlv2_name",
        "tlv2",
        "p5id",
        <Link
          to="/practitioners/edit/p5id"
        >
          Edit
        </Link>,
      ]
    }
  >
    <td
      className="listview-th"
      key="0"
    >
      tlv2_name
    </td>
    <td
      className="listview-th"
      key="1"
    >
      tlv2
    </td>
    <td
      className="listview-th"
      key="2"
    >
      p5id
    </td>
    <td
      className="listview-th"
      key="3"
    >
      <Link
        key="action-p5id"
        to="/practitioners/edit/p5id"
      >
        <LinkAnchor
          href="/practitioners/edit/p5id"
          navigate={[Function]}
        >
          <a
            href="/practitioners/edit/p5id"
            onClick={[Function]}
          >
            Edit
          </a>
        </LinkAnchor>
      </Link>
    </td>
  </ElementMap>
</tr>
`;
